vignettes/s3-vector.Rmd

```diff
@@ -940,27 +931,223 @@ vec_arith.vctrs_meter.MISSING <- function(op, x, y, ...) {
+
+The greatest boilerplate is needed for `vec_ptype2.pizza_percent()` because it is both a method and a generic (thanks to the double-dispatch trick):
+
+* As an S3 method, it implements `vctrs::vec_ptype2()` for the `vctrs_percent`
+  class.
+
+* As a generic, `vec_ptype2.vctrs_percent()` has `default`, `vctrs_percent`, 
+  and `double` methods (see below).
+
+```{r eval = FALSE}
+#' @method vec_ptype2 pizza_percent   # tell roxygen2 its a method
+#' @export                            # export as method
+#' @export vec_ptype2.pizza_percent   # export as generic
+#' @rdname pizza-vctrs                # document (internally)
+vec_ptype2.pizza_percent <- function(x, y, ...) {
+  UseMethod("vec_ptype2.pizza_percent", y) 
+}
+```
+
+(The marginal comments are just for exposition and can be omitted in your own packages.)
+
+This creative take on the S3 system wasn't considered in the design of roxygen2, so we have to spell all of this out explicitly. Since `vec_ptype2.pizza_percent()` is a generic, `R CMD check` will complain if it is not documented, although this documentation is not relevant for users. This is the purpose of the dummy internal help topic defined earlier.
+
+Remember to run `devtools::document()` to re-generate the `NAMESPACE`, and then take a quick look to familiarise yourself with what is generated. You should see this function appearing in two places (if you've never looked at `NAMESPACE` before note that roxygen2 puts it in alphabetical order):
+
+```{r, eval = FALSE}
+S3method(vec_ptype2,pizza_percent)
+...
+export(vec_ptype2.pizza_percent)
 ```
 
-You also need to register the individual double-dispatch methods. Again, this is harder than it should be because roxygen's heuristics aren't quite right. That means you need to describe the `@method` explicitly:
+Now we finish the work around `vec_ptype2()` by providing all needed methods. Each method has both a `@method` and an `@export` directive. These are purely methods so they don't need to be documented in an `.Rd` file.
 
 ```{r}
-#' @method vec_cast.binned double
+#' @method vec_ptype2.vctrs_percent default
+#' @export                                 
+vec_ptype2.vctrs_percent.default <- function(x, y,
+                                             ...,
+                                             x_arg = "x", y_arg = "y") {
+  vec_default_ptype2(x, y, x_arg = x_arg, y_arg = y_arg)
+}
+
+#' @method vec_ptype2.vctrs_percent vctrs_percent
+#' @export                                       
+vec_ptype2.vctrs_percent.vctrs_percent <- function(x, y, ...) new_percent()
+
+#' @method vec_ptype2.vctrs_percent double       
+#' @export                                       
+vec_ptype2.vctrs_percent.double <- function(x, y, ...) double()
+
+#' @method vec_ptype2.double vctrs_percent
 #' @export
-vec_cast.binned.double <- function(x, y, ...) {
+vec_ptype2.double.vctrs_percent <- function(x, y, ...) double()
+```
+
+The exact same treatment applies to `vec_cast()`, which yields:
```

Comment:
I have fallen the trap many times, so I think describing the treatment as exactly the same might cause wrong copy/pasting.

Reply:
Do you think I have the roxygen2 directives wrong? That is all I aimed to address here with the "same treatment" advice.

Reply:
Perhaps introduce with something like this?

> Whereas `vec_ptype2()` and `vec_cast()` methods have slightly different implementations because they dispatch in different order, they use the same pattern of roxygen2 tags.

Reply:
OK yes I have made the wording more precise now 5fdc470614aefb565ee24bef9a09398085cc889c